File: myproj26/main12-high-order-function.py

# 장식자 (Decorators)
def base_10(fn):
    def wrap(x, y):
        return fn(x, y) + 10
    return wrap


# @base_10
# @base_10
def mysum(x, y):
    return x + y


# 장식자를 쓴 것과 같은 효과
mysum = base_10(base_10(mysum))

print(mysum(1, 2))

File: myproj26/main12-high-order-function.js

// function base_10(fn) {
//     function wrap(x, y) {
//         return fn(x, y) + 10;
//     }
//     return wrap;
// }

// function mysum(x, y) {
//     return x + y;
// }

// mysum = base_10(base_10(mysum));

// console.log(mysum(1, 2));

// arrow function 버전
const base_10 = (fn) => (x, y) => fn(x, y) + 10;

mysum = (x, y) => x + y;

mysum = base_10(base_10(mysum));

console.log(mysum(1, 2));

